File: app/src/main/java/code/name/monkey/retromusic/automix/AudioMetadataInjector.kt
/*
 * AudioMetadataInjector.kt
 * ─────────────────────────────────────────────────────────────────────────────
 * Audio metadata injection service for Milla — Android / jaudiotagger edition.
 *
 * Embeds into a freshly downloaded .flac or .m4a file:
 *   • Title, Artist, Album
 *   • BPM
 *   • Cover Art (JPEG)
 *   • Syllable-Synced Lyrics JSON
 *
 * The result is a 100 % self-sufficient offline file: no network call or Room
 * query is required at playback time to display artwork, tags, or lyrics.
 *
 * ┌──────────┬────────────────────────────────────────────────────────────────┐
 * │ Format   │ Tags written                                                   │
 * ├──────────┼────────────────────────────────────────────────────────────────┤
 * │ FLAC     │ VorbisComment: TITLE, ARTIST, ALBUM, BPM                       │
 * │          │   LYRICS          — plain JSON (broad player compat)           │
 * │          │   SYLLABLE_LYRICS — Milla custom field (syllable renderer)     │
 * │          │ FLAC Picture block, type 3 (front cover)                       │
 * ├──────────┼────────────────────────────────────────────────────────────────┤
 * │ M4A/MP4  │ MP4 atoms: ©nam, ©ART, ©alb, tmpo, ©lyr (plain JSON)          │
 * │          │   ----:com.apple.iTunes:SYLLABLE_LYRICS (JSON free-form atom)  │
 * │          │ covr atom (JPEG)                                               │
 * └──────────┴────────────────────────────────────────────────────────────────┘
 *
 * Lyrics JSON canonical format:
 *   [{"time": 4210, "text": "Primera línea"}, ...]
 *   time = milliseconds from start, text = lyric line / syllable
 *
 * Deezer LYRICS_SYNC_JSON is auto-converted via normalizeDeezerLyricsJson().
 *
 * Dependency (already in app/build.gradle.kts):
 *   implementation(libs.jaudiotagger)
 */
package code.name.monkey.retromusic.automix

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.jaudiotagger.audio.AudioFileIO
import org.jaudiotagger.tag.FieldKey
import org.jaudiotagger.tag.Tag
import org.jaudiotagger.tag.flac.FlacTag
import org.jaudiotagger.tag.images.AndroidArtwork
import org.jaudiotagger.tag.images.PictureTypes
import org.jaudiotagger.tag.mp4.Mp4Tag
import org.jaudiotagger.tag.mp4.field.Mp4TagReverseDnsField
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

object AudioMetadataInjector {

    private const val TAG = "AudioMetadataInjector"

    // ── Tag key constants ──────────────────────────────────────────────────────

    /** Custom VorbisComment key written to FLAC files. */
    private const val VORBIS_SYLLABLE_KEY = "SYLLABLE_LYRICS"

    /** iTunes reverse-DNS namespace for the M4A free-form atom. */
    private const val M4A_ITUNES_DOMAIN = "com.apple.iTunes"

    /** Atom name inside the reverse-DNS namespace. */
    private const val M4A_SYLLABLE_KEY = "SYLLABLE_LYRICS"

    // ── Data model ────────────────────────────────────────────────────────────

    /**
     * All metadata to embed into the audio file.
     *
     * @param title             Track title.
     * @param artist            Artist / performer name.
     * @param album             Album / release title.
     * @param bpm               Beats-per-minute (0 = omit tag).
     * @param coverBytes        Raw JPEG bytes for front-cover art.
     * @param syllableLyricsJson  Canonical JSON string:
     *                          `[{"time":<ms>,"text":"<line>"},…]`
     *                          Pass null to skip lyrics injection.
     */
    data class TrackMetadata(
        val title: String = "",
        val artist: String = "",
        val album: String = "",
        val bpm: Int = 0,
        val coverBytes: ByteArray? = null,
        val syllableLyricsJson: String? = null,
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is TrackMetadata) return false
            return title == other.title &&
                artist == other.artist &&
                album == other.album &&
                bpm == other.bpm &&
                coverBytes.contentEquals(other.coverBytes ?: ByteArray(0)) &&
                syllableLyricsJson == other.syllableLyricsJson
        }
        override fun hashCode(): Int = title.hashCode() * 31 + artist.hashCode()
    }

    // ── Public API ────────────────────────────────────────────────────────────

    /**
     * Inject [metadata] into [audioFile] and save it in-place (IO dispatcher).
     *
     * @return `true` on success, `false` on any error (errors are logged).
     */
    suspend fun inject(audioFile: File, metadata: TrackMetadata): Boolean =
        withContext(Dispatchers.IO) {
            require(audioFile.exists()) { "Audio file not found: ${audioFile.path}" }

            Log.d(TAG, "inject() → ${audioFile.name}")
            try {
                val af  = AudioFileIO.read(audioFile)
                val tag = af.tagOrCreateAndSetDefault

                embedTextTags(tag, metadata)
                embedCoverArt(tag, metadata.coverBytes)
                metadata.syllableLyricsJson?.let { injectLyrics(tag, it, audioFile.extension) }

                AudioFileIO.write(af)
                Log.d(TAG, "✅ Tags written: ${audioFile.name}")
                true
            } catch (e: Exception) {
                Log.e(TAG, "inject() failed for ${audioFile.name}: $e")
                false
            }
        }

    // ── Core tag writers ──────────────────────────────────────────────────────

    private fun embedTextTags(tag: Tag, m: TrackMetadata) {
        if (m.title.isNotBlank())  tag.setField(FieldKey.TITLE,  m.title)
        if (m.artist.isNotBlank()) tag.setField(FieldKey.ARTIST, m.artist)
        if (m.album.isNotBlank())  tag.setField(FieldKey.ALBUM,  m.album)
        if (m.bpm > 0)             tag.setField(FieldKey.BPM,    m.bpm.toString())
        Log.d(TAG, "  Text tags: title='${m.title}' artist='${m.artist}' bpm=${m.bpm}")
    }

    private fun embedCoverArt(tag: Tag, coverBytes: ByteArray?) {
        coverBytes ?: return
        val artwork = AndroidArtwork().apply {
            binaryData  = coverBytes
            mimeType    = "image/jpeg"
            pictureType = PictureTypes.DEFAULT_ID   // 3 = front cover
        }
        runCatching {
            tag.deleteArtworkField()
            tag.setField(artwork)
            Log.d(TAG, "  Cover art embedded (${coverBytes.size} bytes)")
        }.onFailure {
            Log.e(TAG, "  embedCoverArt failed: $it")
        }
    }

    // ── Format-specific lyrics injection ──────────────────────────────────────

    private fun injectLyrics(tag: Tag, lyricsJson: String, ext: String) {
        when (tag) {
            is FlacTag -> injectFlacLyrics(tag, lyricsJson)
            is Mp4Tag  -> injectM4aLyrics(tag, lyricsJson)
            else -> runCatching {
                tag.setField(FieldKey.LYRICS, lyricsJson)
                Log.w(TAG, "  Lyrics stored in generic LYRICS field (ext=$ext)")
            }
        }
    }

    /**
     * FLAC: write two VorbisComment fields.
     *
     *  • `LYRICS`          — read by most external players.
     *  • `SYLLABLE_LYRICS` — read exclusively by Milla's syllable renderer.
     */
    private fun injectFlacLyrics(tag: FlacTag, lyricsJson: String) {
        // Standard LYRICS
        runCatching { tag.setField(FieldKey.LYRICS, lyricsJson) }
            .onFailure { Log.w(TAG, "  FLAC LYRICS field error: $it") }

        // Custom SYLLABLE_LYRICS
        runCatching {
            val field = tag.createField(VORBIS_SYLLABLE_KEY, lyricsJson)
            tag.addField(field)
        }.onFailure { Log.w(TAG, "  FLAC SYLLABLE_LYRICS field error: $it") }

        Log.d(TAG, "  [FLAC] LYRICS + SYLLABLE_LYRICS written (${lyricsJson.length} chars)")
    }

    /**
     * M4A: write two atoms.
     *
     *  • `©lyr`                                       — standard iTunes lyrics.
     *  • `----:com.apple.iTunes:SYLLABLE_LYRICS`      — Milla syllable JSON.
     */
    private fun injectM4aLyrics(tag: Mp4Tag, lyricsJson: String) {
        // Standard ©lyr
        runCatching { tag.setField(FieldKey.LYRICS, lyricsJson) }
            .onFailure { Log.w(TAG, "  M4A ©lyr error: $it") }

        // iTunes free-form reverse-DNS atom
        runCatching {
            val atom = Mp4TagReverseDnsField(
                Mp4TagReverseDnsField.IDENTIFIER,
                M4A_ITUNES_DOMAIN,
                M4A_SYLLABLE_KEY,
                lyricsJson,
            )
            tag.addField(atom)
        }.onFailure { Log.w(TAG, "  M4A SYLLABLE_LYRICS atom error: $it") }

        Log.d(TAG, "  [M4A] ©lyr + $M4A_ITUNES_DOMAIN:$M4A_SYLLABLE_KEY written (${lyricsJson.length} chars)")
    }

    // ── Lyrics JSON helpers ───────────────────────────────────────────────────

    /**
     * Convert Deezer `LYRICS_SYNC_JSON` format to Milla's canonical format.
     *
     * Deezer:   `[{"lrc_timestamp":"[00:04.21]","line":"text"}, …]`
     * Canonical:`[{"time":4210,"text":"text"}, …]`
     *
     * If the input is already canonical (has "time" key), it is returned as-is.
     */
    fun normalizeDeezerLyricsJson(deezerSyncJson: String): String {
        return try {
            val arr   = JSONArray(deezerSyncJson)
            val first = arr.optJSONObject(0) ?: return deezerSyncJson
            if (first.has("time")) return deezerSyncJson   // already canonical

            val out = JSONArray()
            for (i in 0 until arr.length()) {
                val entry = arr.getJSONObject(i)
                val ts    = entry.optString("lrc_timestamp", "").trimStart('[').trimEnd(']')
                val ms    = parseLrcToMs(ts)
                val text  = entry.optString("line", "")
                out.put(JSONObject().apply {
                    put("time", ms)
                    put("text", text)
                })
            }
            out.toString()
        } catch (e: Exception) {
            Log.w(TAG, "normalizeDeezerLyricsJson parse error: $e")
            deezerSyncJson
        }
    }

    /**
     * Convert Tidal LRC-style string lyrics to Milla's canonical JSON.
     *
     * Tidal SYLT text field contains lines like `[mm:ss.xx] text`.
     * Returns canonical JSON array string.
     */
    fun normalizeTidalLyricsJson(tidalLyricsText: String): String {
        return try {
            val out   = JSONArray()
            val regex = Regex("""^\[(\d+:\d+\.\d+)]\s*(.*)$""")
            tidalLyricsText.lines().forEach { line ->
                val match = regex.matchEntire(line.trim()) ?: return@forEach
                val ms    = parseLrcToMs(match.groupValues[1])
                val text  = match.groupValues[2]
                out.put(JSONObject().apply {
                    put("time", ms)
                    put("text", text)
                })
            }
            out.toString()
        } catch (e: Exception) {
            Log.w(TAG, "normalizeTidalLyricsJson parse error: $e")
            "[]"
        }
    }

    /** Parse `mm:ss.xx` or `mm:ss.xxx` → milliseconds (Long). */
    private fun parseLrcToMs(ts: String): Long {
        return try {
            val (mm, ss) = ts.split(":")
            mm.toLong() * 60_000L + (ss.toDouble() * 1_000L).toLong()
        } catch (e: Exception) {
            0L
        }
    }

    /**
     * Read the `SYLLABLE_LYRICS` field back from a tagged FLAC file.
     * Returns null if the field is absent.
     */
    fun readSyllableLyricsFromFlac(audioFile: File): String? {
        return try {
            val af  = AudioFileIO.read(audioFile)
            val tag = af.tag as? FlacTag ?: return null
            tag.getFirst(VORBIS_SYLLABLE_KEY).takeIf { it.isNotBlank() }
        } catch (e: Exception) {
            Log.w(TAG, "readSyllableLyricsFromFlac error: $e")
            null
        }
    }

    /**
     * Read the `SYLLABLE_LYRICS` free-form atom back from a tagged M4A file.
     * Returns null if the atom is absent.
     */
    fun readSyllableLyricsFromM4a(audioFile: File): String? {
        return try {
            val af  = AudioFileIO.read(audioFile)
            val tag = af.tag as? Mp4Tag ?: return null
            val key = "----:$M4A_ITUNES_DOMAIN:$M4A_SYLLABLE_KEY"
            tag.getFirst(key).takeIf { it.isNotBlank() }
        } catch (e: Exception) {
            Log.w(TAG, "readSyllableLyricsFromM4a error: $e")
            null
        }
    }
}

File: audio_metadata_injector.py
"""
audio_metadata_injector.py
━━━━━━━━━━━━━━━━━━━━━━━━━
Standalone audio metadata injection service for Milla.

Embeds into any downloaded audio file (.flac, .m4a, .mp3):
  • Cover Art (JPEG, front-cover picture block)
  • Title, Artist, Album
  • BPM
  • Syllable-Synced Lyrics JSON  ← key feature

The injected file is 100 % self-sufficient for offline playback —
no network call or database query is needed for artwork, tags, or
animated lyrics.

Format → tag mapping
─────────────────────────────────────────────────────────────────────
FLAC  │ VorbisComment:  TITLE, ARTIST, ALBUM, BPM, LYRICS (plain)
      │                 SYLLABLE_LYRICS (Milla custom JSON field)
      │ FLAC Picture block (type 3 = front cover)
──────┼──────────────────────────────────────────────────────────────
M4A   │ MP4 atoms:  ©nam, ©ART, ©alb, tmpo, ©lyr (plain)
      │             ----:com.apple.iTunes:SYLLABLE_LYRICS (JSON)
      │             covr (JPEG)
──────┼──────────────────────────────────────────────────────────────
MP3   │ ID3v2.4:  TIT2, TPE1, TALB, TBPM, APIC (cover)
      │           USLT (unsync JSON, lang=eng, desc=SYLLABLE_LYRICS)
      │           SYLT (sync tuples extracted from JSON, ms format)
─────────────────────────────────────────────────────────────────────

Lyrics JSON canonical format (list of dicts):
  [{"time": 4210, "text": "Primera línea"}, ...]
  • time  — milliseconds from start
  • text  — lyric line / syllable text

Deezer LYRICS_SYNC_JSON is auto-converted:
  [{"lrc_timestamp": "[00:04.21]", "line": "..."}, ...]

Install:  pip install mutagen
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Union

# ── mutagen ──────────────────────────────────────────────────────────────────
from mutagen.flac import FLAC, Picture
from mutagen.mp4 import MP4, MP4Cover, MP4FreeForm
from mutagen.id3 import (
    ID3, ID3NoHeaderError,
    TIT2, TPE1, TALB, TBPM, APIC, USLT, SYLT,
    Encoding, PictureType,
)

# ─────────────────────────────────────────────────────────────────────────────

_LyricsInput = Union[list, str, None]


class AudioMetadataInjector:
    """
    Inject cover art, tags, and syllable-synced lyrics JSON into audio files.

    Usage
    -----
    injector = AudioMetadataInjector()
    injector.inject(
        audio_path   = "Danny Ocean - Corazón.flac",
        cover_bytes  = jpeg_bytes,
        lyrics_json  = lyrics_list_or_json_string,
        title        = "Corazón",
        artist       = "Danny Ocean",
        album        = "No Lo Había Dicho",
        bpm          = 95,
    )
    """

    # Custom iTunes reverse-DNS key used in M4A files
    _M4A_SYLLABLE_ATOM = "----:com.apple.iTunes:SYLLABLE_LYRICS"

    # ── public API ────────────────────────────────────────────────────────────

    def inject(
        self,
        audio_path: Union[str, Path],
        cover_bytes: bytes | None = None,
        lyrics_json: _LyricsInput = None,
        title: str = "",
        artist: str = "",
        album: str = "",
        bpm: int = 0,
    ) -> None:
        """
        Main entry point.  Detects format from extension and dispatches.

        Parameters
        ----------
        audio_path   : Path to the target audio file.
        cover_bytes  : Raw JPEG bytes for front-cover art.
        lyrics_json  : Syllable-synced lyrics — list[dict] or JSON string.
                       Accepts canonical {"time": ms, "text": str} OR
                       Deezer {"lrc_timestamp": "[mm:ss.xx]", "line": str}.
        title        : Track title.
        artist       : Artist name.
        album        : Album / release title.
        bpm          : Beats-per-minute (0 = omit tag).
        """
        path = Path(audio_path)
        if not path.is_file():
            raise FileNotFoundError(f"Audio file not found: {path}")

        ext = path.suffix.lower().lstrip(".")
        lyrics_str = self._normalize_lyrics(lyrics_json)

        print(f"[MetaInjector] Injecting → {path.name}  (format: {ext})")

        dispatch = {
            "flac":  self._inject_flac,
            "m4a":   self._inject_m4a,
            "mp4":   self._inject_m4a,
            "aac":   self._inject_m4a,
            "mp3":   self._inject_mp3,
        }
        handler = dispatch.get(ext)
        if handler is None:
            print(f"[MetaInjector] ⚠  Unsupported format '{ext}' — skipping.")
            return

        handler(path, cover_bytes, lyrics_str, title, artist, album, bpm)
        print(f"[MetaInjector] ✅ Done: {path.name}")

    # ── FLAC ──────────────────────────────────────────────────────────────────

    def _inject_flac(
        self, path: Path,
        cover_bytes: bytes | None, lyrics_str: str | None,
        title: str, artist: str, album: str, bpm: int,
    ) -> None:
        audio = FLAC(str(path))

        # VorbisComment text tags
        if title:   audio["title"]  = title
        if artist:  audio["artist"] = artist
        if album:   audio["album"]  = album
        if bpm > 0: audio["bpm"]    = str(bpm)

        # Standard LYRICS field (broad player compatibility)
        if lyrics_str:
            audio["lyrics"] = lyrics_str
            # Custom SYLLABLE_LYRICS field read by Milla's syllable renderer
            audio["syllable_lyrics"] = lyrics_str
            print(f"  [FLAC] LYRICS + SYLLABLE_LYRICS set ({len(lyrics_str):,} chars)")

        # Cover art — FLAC Picture block (type 3 = front cover)
        if cover_bytes:
            pic = Picture()
            pic.type = PictureType.COVER_FRONT
            pic.mime = "image/jpeg"
            pic.desc = "Cover"
            pic.data = cover_bytes
            audio.clear_pictures()
            audio.add_picture(pic)
            print(f"  [FLAC] Cover art embedded ({len(cover_bytes):,} bytes)")

        audio.save()
        print("  [FLAC] VorbisComment tags saved.")

    # ── M4A / MP4 ─────────────────────────────────────────────────────────────

    def _inject_m4a(
        self, path: Path,
        cover_bytes: bytes | None, lyrics_str: str | None,
        title: str, artist: str, album: str, bpm: int,
    ) -> None:
        audio = MP4(str(path))

        # Standard MP4 atoms
        if title:   audio["\xa9nam"] = [title]     # ©nam — title
        if artist:  audio["\xa9ART"] = [artist]    # ©ART — artist
        if album:   audio["\xa9alb"] = [album]     # ©alb — album
        if bpm > 0: audio["tmpo"]    = [bpm]       # tmpo — BPM (integer atom)

        if lyrics_str:
            # ©lyr — standard iTunes lyrics (plain text, widely read)
            audio["\xa9lyr"] = [lyrics_str]
            # ----:com.apple.iTunes:SYLLABLE_LYRICS — JSON for Milla's renderer
            audio[self._M4A_SYLLABLE_ATOM] = [
                MP4FreeForm(lyrics_str.encode("utf-8"))
            ]
            print(f"  [M4A] ©lyr + SYLLABLE_LYRICS atom set ({len(lyrics_str):,} chars)")

        # Cover art — covr atom (JPEG)
        if cover_bytes:
            audio["covr"] = [MP4Cover(cover_bytes, imageformat=MP4Cover.FORMAT_JPEG)]
            print(f"  [M4A] Cover art embedded ({len(cover_bytes):,} bytes)")

        audio.save()
        print("  [M4A] MP4 atoms saved.")

    # ── MP3 / ID3v2 ───────────────────────────────────────────────────────────

    def _inject_mp3(
        self, path: Path,
        cover_bytes: bytes | None, lyrics_str: str | None,
        title: str, artist: str, album: str, bpm: int,
    ) -> None:
        try:
            audio = ID3(str(path))
        except ID3NoHeaderError:
            audio = ID3()

        if title:   audio.add(TIT2(encoding=Encoding.UTF8, text=title))
        if artist:  audio.add(TPE1(encoding=Encoding.UTF8, text=artist))
        if album:   audio.add(TALB(encoding=Encoding.UTF8, text=album))
        if bpm > 0: audio.add(TBPM(encoding=Encoding.UTF8, text=str(bpm)))

        if lyrics_str:
            # USLT — unsynchronized lyrics frame; stores the raw JSON verbatim
            audio.add(USLT(
                encoding=Encoding.UTF8,
                lang="eng",
                desc="SYLLABLE_LYRICS",
                text=lyrics_str,
            ))

            # SYLT — synchronized lyrics frame; (text, timestamp_ms) tuples
            sylt_entries = self._json_to_sylt_entries(lyrics_str)
            if sylt_entries:
                audio.add(SYLT(
                    encoding=Encoding.UTF8,
                    lang="eng",
                    format=2,           # 2 = milliseconds
                    type=1,             # 1 = lyrics
                    desc="SYLLABLE_SYNC",
                    text=sylt_entries,
                ))
                print(f"  [MP3] SYLT set ({len(sylt_entries)} entries)")

            print(f"  [MP3] USLT set ({len(lyrics_str):,} chars)")

        # Cover art — APIC frame (type 3 = front cover)
        if cover_bytes:
            audio.add(APIC(
                encoding=Encoding.UTF8,
                mime="image/jpeg",
                type=PictureType.COVER_FRONT,
                desc="Cover",
                data=cover_bytes,
            ))
            print(f"  [MP3] Cover art embedded ({len(cover_bytes):,} bytes)")

        audio.save(str(path))
        print("  [MP3] ID3v2.4 tags saved.")

    # ── Helpers ───────────────────────────────────────────────────────────────

    @staticmethod
    def _normalize_lyrics(raw: _LyricsInput) -> str | None:
        """
        Coerce any lyrics input to a canonical JSON string.

        Canonical format:  [{"time": <int ms>, "text": "<line>"}, ...]

        Accepts
        -------
        None                                  → None
        str                                   → returned as-is
        list[{"time": int, "text": str}]      → JSON-encoded directly
        list[{"lrc_timestamp": str, "line": str}]  ← Deezer format, converted
        """
        if raw is None:
            return None
        if isinstance(raw, str):
            return raw

        normalized: list[dict] = []
        for entry in raw:
            if "time" in entry:
                # Already canonical
                normalized.append({"time": int(entry["time"]), "text": str(entry.get("text", ""))})
            elif "lrc_timestamp" in entry:
                # Deezer: "[mm:ss.xx]"
                ts_raw = entry.get("lrc_timestamp", "").strip("[]")
                try:
                    mm, ss = ts_raw.split(":")
                    ms = int(float(mm) * 60_000 + float(ss) * 1_000)
                except Exception:
                    ms = 0
                normalized.append({"time": ms, "text": str(entry.get("line", ""))})

        return json.dumps(normalized, ensure_ascii=False)

    @staticmethod
    def _json_to_sylt_entries(lyrics_str: str) -> list[tuple[str, int]]:
        """Convert canonical JSON string to SYLT (text, timestamp_ms) tuples."""
        try:
            entries = json.loads(lyrics_str)
            return [
                (str(e.get("text", "")), int(e.get("time", 0)))
                for e in entries
                if "time" in e
            ]
        except Exception:
            return []
